README

Assuming you've built Gaim from source:

	./autogen.sh && make all
	su -c "make install"

and eveything should be fine. If Gaim is installed from an RPM, then:

	./autogen.sh && make dist
	su -c "rpmbuild -tb gaim-meanwhile-*.tar.gz"

should create binary RPMs. Installing those should get Meanwhile working.



ChangeLog

version 1.2.8 (2006-01-20)

	- fix #1398218 Crash when receiving invite with blank details

	- fix #1403224 Empty message in group chat causes crash
	
version 1.2.7 (2005-12-29)

	- fix #1382062 Angled brackets (<text>) in status break gaim

	- fix #1385556 merge buddy list crash

	- fix #1380778 Logging into meanwhile causes gaim to crash
	
version 1.2.6 (2005-12-13)

	- building against meanwhile 1.0.0
	
	- fix bug when sending mime (image) messages to notesbuddy

	- fix #1354344 status message not changed when idle
	
	- another attempt at fixing #1208464 Charset selection

	- fixes to buddy list synch and merge options

	- rfe #1365378 Set Client ID run-time (added "Hide Client
	Identity" option to account preferences)

	- decode unicode entity values for logging
	
version 1.2.5 (2005-10-28)

	- building against meanwhile 0.5.0

	- fix #1208464 Charset selection

	- rfe #1276721 Support for External buddies

	- support for DH RC2/128 channel and login encryption

	- re-introduced support for idle times

	- rfe #1307892 Support for Announcements

	- fix for crash in adding multi-resolved buddies

	- added "Synchronize with Server" buddy list option

	- fix #1218284 meanwhile ignores ICT chat invites

	- moved "Force Login (Ignore Login Redirects)" preference to
	account options

	- removed "Save NAB group members locally" preference, defaulted
	to TRUE

version 1.2.4 (2004-07-14)

	- fix #1220550 gaim-meanwhile hangs when redirecting to Sametime
	server
	
	- fix problem parsing 8bit parts in MIME messages

	- fix #1232922 status not applied on auto-reconnect
	
	- fix #1120347 Win32 gaim-meanwhile disconnects (error in read
	callback)
	
	- rfe #1199253 Aliases clobber with Notes Address Book (added
	option to not mark NAB group buddies as transient, so that they'll
	save between Gaim sessions)

	- fix tooltip rendering bug for status messages with tag-like text
	
version 1.2.3 (2005-05-27)
	
	- building against meanwhile 0.4.2

	- rfe #1199253 Aliases clobber with Notes Address Book (added
	resolve hook to end of buddy add from NAB groups)
	
	- rfe #1054813 privacy support (allow all, allow list, deny all,
	deny list, allow buddy list)

	- fix #1194631 Windows file transfer crashes gaim
	
	- fix #1178603 crash on duplicate ldap entries (add buddy)

	- added "Invite to Conference..." context menu

	- don't reset conversation features when remote closes
	conversation; prevents sending of formatted messages as plaintext.

	- rfe #1203358 release info... (added version/credits to plugin
	preferences pane)
	
version 1.2.2 (2005-05-02)

	- fix #1193994 startup crash, whitespace status messages
	
version 1.2.1 (2005-05-01)

	- building against meanwhile 0.4.1

	- removed username split for Server, moving it back to an account
	setting. This should fix log location the way it was pre-1.2.0

	- support for sending NotesBuddy-style images
	
	- support for sending/receiving files over the File Transfer
	service
	
	- added "Supports" section to tooltip text for buddies, displaying
	A/V and File Transfer capabilities
	
	- fix #1185763 gaim-meanwhile 1.2.0 not showing Notesbuddy
	messages

	- added context menu item for NAB groups
	
	- differentiate between read-callback disconnect codes
	
	- support for adding, loading, saving Notes Address Book groups
	
	- quiet down error messages when the other end of a conversation
	logs off. Only display when there are still unsent messages in the
	conversation queue
	
	- added support for sending the LoginCont message when the server
	issues a LoginRedirect if the redirect connect fails

	- added a plugin option to always LoginCont and never attempt a
	redirected login
	
	- fixed a problem with moving buddies between groups that would
	cause the change not to be saved.

	- changed the active/away/busy messages dialog to only allow
	single-line messages, to prevent problems when the messages are
	saved to the storage service (and possibly loaded by another
	client, which doesn't permit multi-line messages). Custom away
	messages may still be multi-line

	- Get Info for buddies in the blist is functional
	
version 1.2.0 (2005-03-18)

        - building against meanwhile 0.4.0

	- building against gaim 1.2.0

	- move Server option to username split
	
	- associate known client type identifier with buddies and display
	in tooltip information

	- perform resolve lookup on single-buddy additions to ensure they
	really exist, and to prompt user if multiple matches exist
	
	- slightly better handling of server_alias data from loading of
	buddy list from server

	- patch #1076834 Away/available state and debug logging

	- 'psychic mode' feature, to open an IM window when someone opens
	an IM to you (before they send a message)

	- protocol action now permits setting active/away/busy messages
	
	- load and save active/away/busy messages to storage service

	- support for sending/receiving NotesBuddy HTML

	- support for receiving NotesBuddy images
	
version 1.0.2 (2004-11-09)
	
	- fix #1054818 html/xml tags in group chats
	
	- fix a really stupid bug preventing the adding of new buddies.

	- adds protocol action to set active message

version 1.0.1 (2004-10-15)

	- fix #1045727 gaim crash in mwgaim.c:mw_read_callback after
	disconnect

	- fix #1043849 CPU at 100% on reconnect
	
	- patch #1034727 Add mw_chat_name

	- patch #1044218 Gaim_normalize called for meanwhile: add
	mw_normalize

version 1.0.0 (2004-09-18)

	- version change to match Gaim, because that's what we do

	- g_log calls for meanwhile and mwgaim no longer result in a DOS
	console under win32

version 0.82 (2004-08-26)
	
	- updates to prpl structure as per Gaim API change

	- joining a conference no longer shows a "joined" message for
	existing conference members

	- receiving html-like text is now displayed correctly verbatim

	- added support for compiling using a linux/mingw cross compiler
	
version 0.81 (2004-08-06)

	- Changed away status handling to strip the HTML from the away

src/mime.h

/*
  Meanwhile Protocol Plugin for Gaim
  Adds Lotus Sametime support to Gaim using the Meanwhile library

  Copyright (C) 2004 Christopher (siege) O'Brien <siege@preoccupied.net>
  
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or (at
  your option) any later version.
  
  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
  USA.
*/

#ifndef _GAIM_MIME_H
#define _GAIM_MIME_H


#include <glib.h>
#include <glib/glist.h>


/** @file mime.h

    Rudimentary parsing of multi-part MIME messages into more
    accessible structures.

    Hopefully this might get included into Gaim directly at some
    point. But until then it's just a piece of the Meanwhile plugin.
*/


/** @typedef GaimMimeDocument
    A MIME document
 */
typedef struct _GaimMimeDocument GaimMimeDocument;


/** @typedef GaimMimePart
    A part of a multipart MIME document
 */
typedef struct _GaimMimePart GaimMimePart;


/** Allocate an empty MIME document */
GaimMimeDocument *gaim_mime_document_new();


/** Frees memory used in a MIME document and all of its parts and fields
    @param doc   the MIME document to free
 */
void gaim_mime_document_free(GaimMimeDocument *doc);


/** Parse a MIME document from a NUL-terminated string
    @param buf  the NULL-terminated string containing the MIME-encoded data
    @returns MIME document
 */
GaimMimeDocument *gaim_mime_document_parse(const char *buf);


/** Parse a MIME document from a string
    @param buf  the string containing the MIME-encoded data
    @param len  length of buf
    @returns MIME document
 */
GaimMimeDocument *gaim_mime_document_parsen(const char *buf, gsize len);


/** Write (append) a MIME document onto a GString */
void gaim_mime_document_write(GaimMimeDocument *doc, GString *str);


/** The list of fields in the header of a document

    @param doc  the MIME document
    @returns    list of strings indicating the fields (but not the values of
                the fields) in the header of doc
*/
const GList *gaim_mime_document_get_fields(GaimMimeDocument *doc);


/** Get the value of a specific field in the header of a document

    @param doc    the MIME document
    @param field  case-insensitive field name
    @returns      value associated with the indicated header field, or
                  NULL if the field doesn't exist
*/
const char *gaim_mime_document_get_field(GaimMimeDocument *doc,
					 const char *field);


/** Set or replace the value of a specific field in the header of a
    document

    @param doc    the MIME document
    @param field  case-insensitive field name
    @param value  value to associate with the indicated header field,
                  of NULL to remove the field
*/
void gaim_mime_document_set_field(GaimMimeDocument *doc,
				  const char *field,
				  const char *value);


/** The list of parts in a multipart document

    @param doc   the MIME document
    @returns     list of GaimMimePart contained within doc
*/
const GList *gaim_mime_document_get_parts(GaimMimeDocument *doc);


/** Create and insert a new part into a MIME document

    @param doc   the new part's parent MIME document
 */
GaimMimePart *gaim_mime_part_new(GaimMimeDocument *doc);


/** The list of fields in the header of a document part

    @param part  the MIME document part
    @returns  list of strings indicating the fields (but not the values
              of the fields) in the header of part
*/
const GList *gaim_mime_part_get_fields(GaimMimePart *part);


/** Get the value of a specific field in the header of a document part

    @param part   the MIME document part
    @param field  case-insensitive name of the header field
    @returns      value of the specified header field, or NULL if the
                  field doesn't exist
*/
const char *gaim_mime_part_get_field(GaimMimePart *part,
				     const char *field);


/** Get the decoded value of a specific field in the header of a
    document part */
char *gaim_mime_part_get_field_decoded(GaimMimePart *part,
				       const char *field);


/** Set or replace the value of a specific field in the header of a
    document
    @param doc    the MIME document
    @param field  case-insensitive field name
    @param value  value to associate with the indicated header field,
                  of NULL to remove the field
*/
void gaim_mime_part_set_field(GaimMimePart *part,
			      const char *field,
			      const char *value);


/** Get the (possibly encoded) data portion of a MIME document part
    @param part   the MIME document part
    @returns      NULL-terminated data found in the document part
 */
const char *gaim_mime_part_get_data(GaimMimePart *part);


/** Get the data portion of a MIME document part, after attempting to
    decode it according to the content-transfer-encoding field. If the
    specified encoding method is not supported, this function will
    return NULL.

    @param part the MIME documemt part
    @param 
*/
void gaim_mime_part_get_data_decoded(GaimMimePart *part,
				     char **data, gsize *len);


/** Get the length of the data portion of a MIME document part

    @param part  the MIME document part
    @returns     length of the data in the document part
*/
gsize gaim_mime_part_get_length(GaimMimePart *part);


void gaim_mime_part_set_data(GaimMimePart *part, const char *data);


#endif


src/mime.c

/*
  Meanwhile Protocol Plugin for Gaim
  Adds Lotus Sametime support to Gaim using the Meanwhile library

  Copyright (C) 2004 Christopher (siege) O'Brien <siege@preoccupied.net>
  
  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or (at
  your option) any later version.
  
  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
  USA.
*/

/* this file becomes part of core as of Gaim 2.0.0 */

#include <stdio.h>
#include <string.h>

#include <glib.h>
#include <glib/ghash.h>
#include <glib/glist.h>
#include <glib/gstring.h>

#include <debug.h>
#include <util.h>

#include "mime.h"


/** @struct mime_fields

    Utility structure used in both MIME document and parts, which maps 
    field names to their values, and keeps an easily accessible list of
    keys
*/
struct mime_fields {
  GHashTable *map;
  GList *keys;
};


struct _GaimMimeDocument {
  struct mime_fields fields;
  GList *parts;
};


struct _GaimMimePart {
  struct mime_fields fields;
  struct _GaimMimeDocument *doc;
  GString *data;
};


static void fields_set(struct mime_fields *mf,
		       const char *key, const char *val) {
  char *k, *v;

  g_return_if_fail(mf != NULL);
  g_return_if_fail(mf->map != NULL);

  k = g_ascii_strdown(key, -1);
  v = g_strdup(val);

  /* append to the keys list only if it's not already there */
  if(! g_hash_table_lookup(mf->map, k)) {
    mf->keys = g_list_append(mf->keys, k);
  }

  /* important to use insert. If the key is already in the table, then
     it's already in the keys list. Insert will free the new instance
     of the key rather than the old instance. */
  g_hash_table_insert(mf->map, k, v);
}


static const char *fields_get(struct mime_fields *mf,
			      const char *key) {
  char *kdown;
  const char *ret;

  g_return_val_if_fail(mf != NULL, NULL);
  g_return_val_if_fail(mf->map != NULL, NULL);

  kdown = g_ascii_strdown(key, -1);
  ret = g_hash_table_lookup(mf->map, kdown);
  g_free(kdown);

  return ret;
}


static void fields_init(struct mime_fields *mf) {
  g_return_if_fail(mf != NULL);

  mf->map = g_hash_table_new_full(g_str_hash, g_str_equal,
				  g_free, g_free);
}


static void fields_loadline(struct mime_fields *mf,
			    const char *line, gsize len) {

  /* split the line into key: value */
  char *key, *val;
  char **tokens;

  /* feh, need it to be NUL terminated */
  key = g_strndup(line, len);

  /* split */
  val = strchr(key, ':');
  if(! val) {
    g_free(key);
    return;
  }
  *val++ = '\0';

  /* normalize whitespace (sorta) and trim on key and value */
  tokens = g_strsplit(key, "\t\r\n", 0);
  key = g_strjoinv("", tokens);
  key = g_strstrip(key);
  g_strfreev(tokens);

  tokens = g_strsplit(val, "\t\r\n", 0);
  val = g_strjoinv("", tokens);
  val = g_strstrip(val);
  g_strfreev(tokens);
  
  fields_set(mf, key, val);

  g_free(key);
  g_free(val);
}


static void fields_load(struct mime_fields *mf,
			char **buf, gsize *len) {
  char *tail;

  while( (tail = g_strstr_len(*buf, *len, "\r\n")) ) {
    char *line;
    gsize ln;

    /* determine the current line */
    line = *buf;
    ln = tail - line;

    /* advance our search space past the CRLF */
    *buf = tail + 2;
    *len -= (ln + 2);

    /* empty line, end of headers */
    if(! ln) return;

    /* look out for line continuations */
    if(line[ln-1] == ';') {
      tail = g_strstr_len(*buf, *len, "\r\n");
      if(tail) {
	gsize cln;

	cln = tail - *buf;
	ln = tail - line;

	/* advance our search space past the CRLF (again) */
	*buf = tail + 2;
	*len -= (cln + 2);
      }
    }

    /* process our super-cool line */
    fields_loadline(mf, line, ln);
  }
}


static void field_write(const char *key, const char *val, GString *str) {
  g_string_append_printf(str, "%s: %s\r\n", key, val);
}


static void fields_write(struct mime_fields *mf, GString *str) {
  g_return_if_fail(mf != NULL);

  g_hash_table_foreach(mf->map, (GHFunc) field_write, str);
  g_string_append(str, "\r\n");
}


static void fields_destroy(struct mime_fields *mf) {
  g_return_if_fail(mf != NULL);

  g_hash_table_destroy(mf->map);
  g_list_free(mf->keys);

  mf->map = NULL;
  mf->keys = NULL;
}


static GaimMimePart *part_new(GaimMimeDocument *doc) {
  GaimMimePart *part;
  
  part = g_new0(GaimMimePart, 1);
  fields_init(&part->fields);
  part->doc = doc;
  part->data = g_string_new(NULL);

  doc->parts = g_list_prepend(doc->parts, part);

  return part;
}


static void part_load(GaimMimePart *part,
		      const char *buf, gsize len) {

  char *b = (char *) buf;
  gsize n = len;

  fields_load(&part->fields, &b, &n);

  /* the remainder will have a blank line, if there's anything at all,
     so check if there's anything then trim off the trailing four
     bytes, \r\n\r\n */
  if(n > 4) n -= 4;
  g_string_append_len(part->data, b, n);
}


static void part_write(GaimMimePart *part, GString *str) {
  fields_write(&part->fields, str);
  g_string_append_printf(str, "%s\r\n\r\n", part->data->str);
}


static void part_free(GaimMimePart *part) {

  fields_destroy(&part->fields);

  g_string_free(part->data, TRUE);
  part->data = NULL;

  g_free(part);
}


GaimMimePart *gaim_mime_part_new(GaimMimeDocument *doc) {
  g_return_val_if_fail(doc != NULL, NULL);
  return part_new(doc);
}


const GList *gaim_mime_part_get_fields(GaimMimePart *part) {
  g_return_val_if_fail(part != NULL, NULL);
  return part->fields.keys;
}


const char *gaim_mime_part_get_field(GaimMimePart *part,
				     const char *field) {

  g_return_val_if_fail(part != NULL, NULL);
  return fields_get(&part->fields, field);
}


char *gaim_mime_part_get_field_decoded(GaimMimePart *part,
				       const char *field) {

  const char *f;

  g_return_val_if_fail(part != NULL, NULL);

  f = fields_get(&part->fields, field);
  return gaim_mime_decode_field(f);
}


void gaim_mime_part_set_field(GaimMimePart *part,
			      const char *field,
			      const char *value) {
  g_return_if_fail(part != NULL);
  fields_set(&part->fields, field, value);
}


const char *gaim_mime_part_get_data(GaimMimePart *part) {
  g_return_val_if_fail(part != NULL, NULL);
  g_return_val_if_fail(part->data != NULL, NULL);

  return part->data->str;
}


void gaim_mime_part_get_data_decoded(GaimMimePart *part,
				     char **data, gsize *len) {
  const char *enc;

  g_return_if_fail(part != NULL);
  g_return_if_fail(data != NULL);
  g_return_if_fail(len != NULL);

  g_return_if_fail(part->data != NULL);

  enc = gaim_mime_part_get_field(part, "content-transfer-encoding");

  if(! enc) {
    *data = g_strdup(part->data->str);
    *len = part->data->len;

  } else if(! g_ascii_strcasecmp(enc, "7bit")) {
    *data = g_strdup(part->data->str);
    *len = part->data->len;

  } else if(! g_ascii_strcasecmp(enc, "8bit")) {
    *data = g_strdup(part->data->str);
    *len = part->data->len;

  } else if(! g_ascii_strcasecmp(enc, "base16")) {
    *len = gaim_base16_decode(part->data->str, (unsigned char **) data);

  } else if(! g_ascii_strcasecmp(enc, "base64")) {
    gaim_base64_decode(part->data->str, data, (int *) len);

  } else if(! g_ascii_strcasecmp(enc, "quoted-printable")) {
    gaim_quotedp_decode(part->data->str, data, (int *) len);

  } else {
    gaim_debug_warning("mime", "gaim_mime_part_get_data_decoded:"
		       " unknown encoding '%s'\n", enc);
    *data = NULL;
    *len = 0;
  }
}


gsize gaim_mime_part_get_length(GaimMimePart *part) {
  g_return_val_if_fail(part != NULL, 0);
  g_return_val_if_fail(part->data != NULL, 0);

  return part->data->len;
}


void gaim_mime_part_set_data(GaimMimePart *part, const char *data) {
  g_return_if_fail(part != NULL);
  g_string_free(part->data, TRUE);
  part->data = g_string_new(data);
}


GaimMimeDocument *gaim_mime_document_new() {
  GaimMimeDocument *doc;

  doc = g_new0(GaimMimeDocument, 1);
  fields_init(&doc->fields);

  return doc;
}


static void doc_parts_load(GaimMimeDocument *doc,
			   const char *boundary,
			   const char *buf, gsize len) {

  char *b = (char *) buf;
  gsize n = len;

  const char *bnd;
  gsize bl;

  bnd = g_strdup_printf("--%s", boundary);
  bl = strlen(bnd);

  for(b = g_strstr_len(b, n, bnd); b; ) {
    char *tail;

    /* skip the boundary */
    b += bl;
    n -= bl;

    /* skip the trailing \r\n or -- as well */
    if(n >= 2) {
      b += 2;
      n -= 2;
    }

    /* find the next boundary */
    tail = g_strstr_len(b, n, bnd);

    if(tail) {
      gsize sl;

      sl = tail - b;
      if(sl) {
	GaimMimePart *part = part_new(doc);
	part_load(part, b, sl);
      }
    }

    b = tail;
  }
}


GaimMimeDocument *gaim_mime_document_parsen(const char *buf, gsize len) {

  GaimMimeDocument *doc;

  char *b = (char *) buf;
  gsize n = len;

  g_return_val_if_fail(buf != NULL, NULL);

  doc = gaim_mime_document_new();

  if(! len) return doc;

  fields_load(&doc->fields, &b, &n);

  {
    const char *ct = fields_get(&doc->fields, "content-type");
    if(ct && g_str_has_prefix(ct, "multipart")) {
      char *bd = strrchr(ct, '=');
      if(bd++) {
	doc_parts_load(doc, bd, b, n);
      }      
    }
  }

  return doc;
}


GaimMimeDocument *gaim_mime_document_parse(const char *buf) {
  g_return_val_if_fail(buf != NULL, NULL);
  return gaim_mime_document_parsen(buf, strlen(buf));
}


void gaim_mime_document_write(GaimMimeDocument *doc, GString *str) {
  const char *bd = NULL;

  g_return_if_fail(doc != NULL);
  g_return_if_fail(str != NULL);

  {
    const char *ct = fields_get(&doc->fields, "content-type");
    if(ct && g_str_has_prefix(ct, "multipart")) {
      char *b = strrchr(ct, '=');
      if(b++) bd = b;
    }
  }

  fields_write(&doc->fields, str);

  if(bd) {
    GList *l;

    for(l = doc->parts; l; l = l->next) {
      g_string_append_printf(str, "--%s\r\n", bd);

      part_write(l->data, str);

      if(! l->next) {
	g_string_append_printf(str, "--%s--\r\n", bd);
      }
    }
  }
}


const GList *gaim_mime_document_get_fields(GaimMimeDocument *doc) {
  g_return_val_if_fail(doc != NULL, NULL);
  return doc->fields.keys;
}


const char *gaim_mime_document_get_field(GaimMimeDocument *doc,
					 const char *field) {
  g_return_val_if_fail(doc != NULL, NULL);
  return fields_get(&doc->fields, field);
}


void gaim_mime_document_set_field(GaimMimeDocument *doc,
				  const char *field,
				  const char *value) {
  g_return_if_fail(doc != NULL);
  fields_set(&doc->fields, field, value);
}


const GList *gaim_mime_document_get_parts(GaimMimeDocument *doc) {
  g_return_val_if_fail(doc != NULL, NULL);
  return doc->parts;
}


void gaim_mime_document_free(GaimMimeDocument *doc) {
  if(! doc) return;

  fields_destroy(&doc->fields);

  while(doc->parts) {
    part_free(doc->parts->data);
    doc->parts = g_list_delete_link(doc->parts, doc->parts);
  }

  g_free(doc);
}

